src/source/split_at.rs

```diff
@@ -0,0 +1,123 @@
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+
+use crate::ChannelCount;
+use crate::SampleRate;
+
+use super::Source;
+use super::TrackPosition;
+
+pub struct SplitAt<S> {
+    shared_source: Arc<Mutex<Option<TrackPosition<S>>>>,
+    active: Option<TrackPosition<S>>,
+    split_duration: Option<Duration>,
+    split_point: Duration,
+    first_span_sample_rate: SampleRate,
+    first_span_channel_count: ChannelCount,
+}
+
+impl<S> SplitAt<S>
+where
+    S: Source,
+    <S as Iterator>::Item: crate::Sample,
+{
+    /// returns two sources, the second is inactive and will return
+    /// none until the first has passed the split_point.
+    pub(crate) fn new(input: S, split_point: Duration) -> [Self; 2] {
+        let shared_source = Arc::new(Mutex::new(None));
+        let first_span_sample_rate = input.sample_rate();
+        let first_span_channel_count = input.channels();
+        let total_duration = input.total_duration();
+        [
+            Self {
+                shared_source: shared_source.clone(),
+                active: Some(input.track_position()),
+                split_duration: Some(split_point),
+                split_point,
+                first_span_sample_rate,
+                first_span_channel_count,
```

Comment:
Maybe the second source can be of different type that just awaits until the values are available? These fields are only useful there.
Maybe we should introduce some kind of `UniformSource` that does not have spans, or bite the bullet and document that in some cases we only expect `current_span_len() == None`. The latter is messier but can be done with existing API.

Reply:
> Maybe the second source can be of different type that just awaits until the values are available? These fields are only useful there.

I thought about that however this way we can reuse `SplitAt` for a future function that splits into more then two splits.

> Maybe we should introduce some kind of UniformSource that does not have spans, or bite the bullet and document that in some cases we only expect current_span_len() == None. The latter is messier but can be done with existing API.

`current_span_len().is_none()` already has the meaning that the span will never change. At least thats what the trait doc says I think....

Reply:
I got it: we will make the `current_span_len` report `Some(1)` while the segment is inactive. That should trigger a re-query of sample-rate and channel count every time. 

`Some(0)` is documented to mean no more data so that would not work.